flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteTierProducerAgent.java

```diff
@@ -0,0 +1,110 @@
+
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.common.TieredStoragePartitionId;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.common.TieredStorageSubpartitionId;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.file.PartitionFileWriter;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage.TieredStorageMemoryManager;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage.TieredStorageResourceRegistry;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.tier.TierProducerAgent;
+
+import java.util.Arrays;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/** The implementation of {@link TierProducerAgent} for the remote tier. */
+public class RemoteTierProducerAgent implements TierProducerAgent {
+
+    private final int numSubpartitions;
+
+    private final int numBuffersPerSegment;
+
+    private final RemoteCacheManager cacheDataManager;
+
+    private final int[] currentSubpartitionWriteBuffers;
+
+    RemoteTierProducerAgent(
+            TieredStoragePartitionId partitionId,
+            int numSubpartitions,
+            int numBytesPerSegment,
+            int bufferSizeBytes,
+            boolean isBroadcastOnly,
+            PartitionFileWriter partitionFileWriter,
+            TieredStorageMemoryManager memoryManager,
+            TieredStorageResourceRegistry resourceRegistry) {
+        checkArgument(
+                numBytesPerSegment >= bufferSizeBytes,
+                "One segment should contain at least one buffer.");
+
+        this.numSubpartitions = numSubpartitions;
+        this.numBuffersPerSegment = numBytesPerSegment / bufferSizeBytes;
+        this.cacheDataManager =
+                new RemoteCacheManager(
+                        partitionId,
+                        isBroadcastOnly ? 1 : numSubpartitions,
+                        memoryManager,
+                        partitionFileWriter);
+        this.currentSubpartitionWriteBuffers = new int[numSubpartitions];
+        Arrays.fill(currentSubpartitionWriteBuffers, 0);
+        resourceRegistry.registerResource(partitionId, this::releaseAllResources);
+    }
+
+    @Override
+    public boolean tryStartNewSegment(TieredStorageSubpartitionId subpartitionId, int segmentId) {
+        cacheDataManager.startSegment(subpartitionId.getSubpartitionId(), segmentId);
+        // The remote storage tier should always be able to start a new segment.
+        return true;
+    }
+
+    @Override
+    public boolean tryWrite(TieredStorageSubpartitionId subpartitionId, Buffer buffer) {
+        int subpartitionIndex = subpartitionId.getSubpartitionId();
+        if (currentSubpartitionWriteBuffers[subpartitionIndex] != 0
+                && currentSubpartitionWriteBuffers[subpartitionIndex] + 1 > numBuffersPerSegment) {
```

Comment:
We already checked in the constructor that one segment at least contain one buffer. `currentSubpartitionWriteBuffers[subpartitionIndex] != 0` is probably not needed here.

Reply:
Fixed.